lib/integrations/awsra/generate_credentials.go

```diff
@@ -0,0 +1,233 @@
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/utils/keys"
+	"github.com/gravitational/teleport/lib/cryptosuites"
+	"github.com/gravitational/teleport/lib/integrations/awsra/createsession"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+const (
+	// AWSCredentialsSourceRolesAnywhere is the source of the credentials.
+	// There's no official constant in AWS docs.
+	AWSCredentialsSourceRolesAnywhere = "RolesAnywhere"
+)
+
+// CertificateGenerator is an interface that generates a certificate.
+type CertificateGenerator interface {
+	GenerateCertificate(req tlsca.CertificateRequest) ([]byte, error)
+}
+
+// GenerateCredentialsRequest is a request to generate AWS credentials using the AWS IAM Roles Anywhere integration.
+type GenerateCredentialsRequest struct {
+	// Clock is the clock used to calculate the expiration time of the credentials.
+	Clock clockwork.Clock
+	// TrustAnchorARN is the ARN of the AWS IAM Roles Anywhere Trust Anchor.
+	TrustAnchorARN string
+	// ProfileARN is the ARN of the AWS IAM Roles Anywhere Profile.
+	ProfileARN string
+	// RoleARN is the ARN of the AWS IAM Role to generate credentials.
+	RoleARN string
+	// SubjectCommonName is the common name to use in the certificate.
+	SubjectCommonName string
+	// DurationSeconds is the duration of the session.
+	// If nil, the default duration of the Profile will be used.
+	DurationSeconds *int
+	// AcceptRoleSessionName indicates whether this Profile accepts a role session name.
+	// Setting the role session name when the Profile does not accept it will result in an error.
+	AcceptRoleSessionName bool
+	// KeyStoreManager grants access to the AWS Roles Anywhere signer.
+	KeyStoreManager KeyStoreManager
+	/// Cache is used to get the current cluster name and cert authority keys.
+	Cache Cache
+
+	// CreateSession is the API used to create a session with AWS IAM Roles Anywhere.
+	// This is used to mock the CreateSession API in tests.
+	CreateSession func(ctx context.Context, req createsession.CreateSessionRequest) (*createsession.CreateSessionResponse, error)
+}
+
+// KeyStoreManager defines methods to get signers using the server's keystore.
+type KeyStoreManager interface {
+	// GetTLSCertAndSigner selects a usable TLS keypair from the given CA
+	// and returns the PEM-encoded TLS certificate and a [crypto.Signer].
+	GetTLSCertAndSigner(ctx context.Context, ca types.CertAuthority) ([]byte, crypto.Signer, error)
+}
+
+// Cache is the subset of the cached resources that the Service queries.
+type Cache interface {
+	// GetCertAuthority returns cert authority by id
+	GetCertAuthority(ctx context.Context, id types.CertAuthID, loadKeys bool) (types.CertAuthority, error)
+	// GetClusterName returns the current cluster name.
+	GetClusterName(ctx context.Context) (types.ClusterName, error)
+}
+
+func (r *GenerateCredentialsRequest) checkAndSetDefaults() error {
+	r.Clock = cmp.Or(r.Clock, clockwork.NewRealClock())
```

Comment:
Is there a reason to prefer 
```
r.Clock = cmp.Or(r.Clock, clockwork.NewRealClock())
```
over 
```
if r.Clock == nil {
    r.Clock = clockwork.NewRealClock()
}
```
?

Reply:
I think that `r.Clock = cmp.Or(r.Clock, clockwork.NewRealClock())` always create a real clock to discard it just after entering the function whereas `if r.Clock == nil` doesn't.